crates/ironrdp-client-blocking/src/rdp.rs

```diff
@@ -0,0 +1,269 @@
+
+    debug!("TLS upgrade");
+
+    // Ensure there is no leftover
+    let initial_stream = framed.into_inner_no_leftover();
+
+    let (upgraded_stream, server_public_key) = upgrade_blocking(initial_stream, config.destination.name())
+        .map_err(|e| connector::custom_err!("TLS upgrade", e))?;
+
+    let upgraded = ironrdp_blocking::mark_as_upgraded(should_upgrade, &mut connector, server_public_key);
+
+    let mut upgraded_framed = ironrdp_blocking::Framed::new(upgraded_stream);
+
+    let connection_result = ironrdp_blocking::connect_finalize(upgraded, &mut upgraded_framed, connector)?;
+
+    Ok((connection_result, upgraded_framed))
+}
+
+// TODO: should probably be moved to ironrdp-tls
+fn upgrade_blocking(stream: TcpStream, server_name: &str) -> io::Result<(native_tls::TlsStream<TcpStream>, Vec<u8>)> {
+    let tls_connector =
+        native_tls::TlsConnector::new().map_err(|e| connector::custom_err!("create TlsConnector", e))?;
+    let tls_stream = tls_connector
+        .connect(server_name, stream)
+        .map_err(|e| connector::custom_err!("TLS connect", e))?;
+
+    let server_public_key = {
+        let cert = tls_stream
+            .peer_certificate()
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?
+            .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "peer certificate is missing"))?;
+        let cert = cert.to_der().map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
+        extract_tls_server_public_key(&cert)?
+    };
+
+    Ok((tls_stream, server_public_key))
+}
+
+// TODO: taken from ironrdp-tls, remove x509_cert from Cargo.toml when this is removed
+fn extract_tls_server_public_key(cert: &[u8]) -> io::Result<Vec<u8>> {
+    use x509_cert::der::Decode as _;
+
+    let cert = x509_cert::Certificate::from_der(cert).map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
+
+    let server_public_key = cert
+        .tbs_certificate
+        .subject_public_key_info
+        .subject_public_key
+        .as_bytes()
+        .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "subject public key BIT STRING is not aligned"))?
+        .to_owned();
+
+    Ok(server_public_key)
+}
+
+fn active_session(
+    framed: UpgradedFramed,
+    connection_result: ConnectionResult,
+    event_loop_proxy: &EventLoopProxy<RdpOutputEvent>,
+    input_event_receiver: &mut crossbeam::channel::Receiver<RdpInputEvent>,
+) -> SessionResult<RdpControlFlow> {
+    let mut image = DecodedImage::new(
+        PixelFormat::RgbA32,
+        connection_result.desktop_size.width,
+        connection_result.desktop_size.height,
+    );
+
+    let mut active_stage = ActiveStage::new(connection_result, None);
+
+    let (pdu_tx, pdu_rx) = crossbeam::channel::unbounded();
+    let framed = Arc::new(std::sync::Mutex::new(framed));
+    let framed_clone = framed.clone();
+
+    std::thread::spawn(move || loop {
+        let frame = framed.lock().unwrap().read_pdu();
+        pdu_tx.send(frame).unwrap();
+    });
```

Comment:
We can only `crossbeam::channel::select!` on a `crossbeam::channel`, so we create a new thread here which reads frames in a loop and sends them to the `pdu_rx` channel, which allows us to more or less imitate the logic in `ironrdp-client`

Reply:
This thread probably won't exit gracefully, I think it will panic once the `select!` loop below exits. That can obviously be fixed.